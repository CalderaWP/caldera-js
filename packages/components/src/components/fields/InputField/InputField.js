import {
	parseAttributes,
	fieldClassNames,
	isValidHtml5type,
	toBoolean
} from '../util';

import PropTypes from 'prop-types';
import React from 'react';
import { TextControl, CheckboxControl } from '@wordpress/components';
export const InputField = ({
	label,
	description,
	fieldId,
	placeholder,
	required,
	html5type,
	value,
	onChange,
	onBlur,
	attributes,
	children
}) => {
	const fieldType = isValidHtml5type(html5type) ? html5type : 'text';
	const _attributes = parseAttributes(attributes, fieldType);
	if ('checkbox' === fieldType) {
		const checked = value ? true : false;
		return (
			<CheckboxControl
				id={fieldId}
				checked={checked}
				help={description}
				{..._attributes}
				onChange={onChange}
				label={label}
				onBlur={onBlur}
			/>
		);
	} else {
		return (
			<TextControl
				label={label}
				className={fieldClassNames(fieldType)}
				id={fieldId}
				value={value}
				placeholder={placeholder}
				type={fieldType}
				onChange={onChange}
				onBlur={onBlur}
				help={description}
				{..._attributes}
			/>
		);
	}
};

InputField.propTypes = {
	label: PropTypes.string,
	description: PropTypes.string,
	fieldId: PropTypes.string,
	placeholder: PropTypes.string,
	required: PropTypes.bool,
	html5type: PropTypes.string,
	attributes: PropTypes.object,
	value: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
		PropTypes.array,
		PropTypes.bool
	]),
	onChange: PropTypes.func,
	onBlur: PropTypes.func
};

InputField.defaultProps = {
	onBlur: () => {},
	required: false,
	html5type: 'text'
};
