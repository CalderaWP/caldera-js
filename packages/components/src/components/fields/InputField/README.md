# Input Field

## Example

```jsx
<InputField
		label={'Text Input'}
		html5type={'text'}
		onChange={onChange}
		onBlur={onBlur}
/>
```