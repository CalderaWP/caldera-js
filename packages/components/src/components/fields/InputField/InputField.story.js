import React from 'react';
import { storiesOf } from '@storybook/react';
import { InputField } from './InputField';

let value = 'Hi Roy';
const onChange = newValue => {
	console.log(newValue);
};
const onBlur = event => {};

const STORY_NAME_INPUT = 'InputField';
const STORY_NAME_INPUT_NUMBER = 'InputField as number field';
const STORY_NAME_INPUT_EMAIL = 'InputField as email field';
const STORY_NAME_INPUT_CHECKBOX = 'InputField as checkbox';
storiesOf(STORY_NAME_INPUT, module).add('With no value', () => (
	<InputField
		label={'Text Input'}
		html5type={'text'}
		onChange={onChange}
		onBlur={onBlur}
	/>
));

storiesOf(STORY_NAME_INPUT, module).add('With placeholder', () => (
	<InputField
		label={'Text Input'}
		placeholder={'Hi Roy'}
		html5type={'text'}
		onChange={onChange}
		onBlur={onBlur}
	/>
));

storiesOf(STORY_NAME_INPUT, module).add('With description', () => (
	<InputField
		label={'Text Input With Description'}
		html5type={'text'}
		description={'Text field description'}
		onChange={onChange}
		onBlur={onBlur}
	/>
));
storiesOf(STORY_NAME_INPUT, module).add('With value', () => (
	<InputField
		label={'Text Input'}
		value={value}
		html5type={'text'}
		onChange={onChange}
		onBlur={onBlur}
	/>
));
storiesOf(STORY_NAME_INPUT, module).add('Disbaled', () => (
	<InputField
		label={'Text Input'}
		value={value}
		html5type={'text'}
		onChange={onChange}
		onBlur={onBlur}
		attributes={{
			disabled: true
		}}
	/>
));

storiesOf(STORY_NAME_INPUT_NUMBER, module).add('With attributes', () => (
	<InputField
		label={'Number'}
		html5type={'number'}
		onChange={onChange}
		onBlur={onBlur}
		attributes={{
			min: 40,
			max: 100,
			step: 10
		}}
	/>
));
storiesOf(STORY_NAME_INPUT_EMAIL, module).add('Required', () => (
	<InputField
		label={'Email'}
		html5type={'email'}
		required={true}
		onChange={onChange}
		description={'Minimum 40, Max 100, Step 10'}
		onBlur={onBlur}
	/>
));
storiesOf(STORY_NAME_INPUT_EMAIL, module).add('Multiple', () => (
	<InputField
		label={'Emails'}
		html5type={'email'}
		required={true}
		onChange={onChange}
		description={'Multiples allowed'}
		onBlur={onBlur}
		attributes={{
			multiple: true
		}}
	/>
));

storiesOf(STORY_NAME_INPUT_CHECKBOX, module).add('Checked', () => (
	<InputField
		label={'Option'}
		html5type={'checkbox'}
		value={true}
		onChange={onChange}
		description={'Set to checked'}
	/>
));

storiesOf(STORY_NAME_INPUT_CHECKBOX, module).add('Not Checked', () => (
	<InputField
		label={'Option'}
		html5type={'checkbox'}
		value={false}
		onChange={onChange}
		description={'Not  checked'}
	/>
));

storiesOf(STORY_NAME_INPUT_CHECKBOX, module).add('Disabled', () => (
	<InputField
		label={'Option'}
		html5type={'checkbox'}
		value={false}
		onChange={onChange}
		description={'Not  checked'}
		attributes={{
			disabled: true
		}}
	/>
));
