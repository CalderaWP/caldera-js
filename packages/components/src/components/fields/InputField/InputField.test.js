import * as React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { InputField } from './InputField';

describe('InputField ', () => {
	let onChange;
	let onBlur;

	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});

	it('matches snapshot with all props', () => {
		const component = renderer.create(
			<InputField
				label={'Hi Roy'}
				description={'Say Hi'}
				placeholder={'Hello'}
				html5type={'text'}
				value={'Roy'}
				onChange={onChange}
				onBlur={onBlur}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Uses text when html5type is not valid', () => {
		const component = mount(
			<InputField
				label={'Hi Roy'}
				description={'Say Hi'}
				placeholder={'Hello'}
				html5type={'unreal'}
				value={'Roy'}
				onChange={onChange}
				onBlur={onBlur}
			/>
		);

		expect(component.find('input').prop('type')).toEqual('text');
	});

	it('Uses html5type as type when html5type is valid', () => {
		const component = mount(
			<InputField
				label={'Hi Roy'}
				description={'Say Hi'}
				placeholder={'Hello'}
				html5type={'email'}
				value={'Roy'}
				onChange={onChange}
				onBlur={onBlur}
			/>
		);

		expect(component.find('input').prop('type')).toEqual('email');
	});

	it('Changes calls change handler', () => {
		const component = mount(
			<InputField
				label={'Hi Roy'}
				description={'Say Hi'}
				placeholder={'Hello'}
				html5type={'email'}
				value={'Roy'}
				onChange={onChange}
				onBlur={onBlur}
			/>
		);

		component.find('input').simulate('change');
		expect(onChange.mock.calls.length).toBe(1);
	});
});
