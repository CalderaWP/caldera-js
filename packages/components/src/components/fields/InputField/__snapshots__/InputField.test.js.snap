// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField  matches snapshot with all props 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <label>
    Hi Roy
     
  </label>
  <input
    aria-describedby="undefined__help"
    className="caldera-field caldera-field-text"
    onBlur={[MockFunction]}
    onChange={[Function]}
    placeholder="Hello"
    type="text"
    value="Roy"
  />
  <p
    className="caldera-control__help"
    id="undefined__help"
  >
    Say Hi
  </p>
</div>
`;
